lib/src/rsa.nr

```diff
@@ -2,8 +2,74 @@ use dep::bignum::BigNum;
 use dep::bignum::runtime_bignum::BigNumInstance;
 use dep::bignum::runtime_bignum::BigNumTrait;
 use dep::bignum::runtime_bignum::BigNumInstanceTrait;
+use dep::bignum::runtime_bignum::BigNumParamsTrait;
+use crate::types::{RSA, BN1024, BN2048, BNInst1024, BNInst2048, RSA1024, RSA2048, BN1964, BNInst1964, RSA1964};
 
-use crate::types::{RSA, BN1024, BN2048, BNInst1024, BNInst2048, RSA1024, RSA2048};
+global HASH_LEN: u32 = 32;
+
+fn reverse_array<let N: u32>(array: [u8; N]) -> [u8; N] {
+    let mut reversed = [0 as u8; N];
+    for i in 0..N {
+        reversed[i] = array[N - i - 1];
+    }
+    reversed
+}
+
+fn get_array_slice<let N: u32, let M: u32>(array: [u8; N], start: u32, end: u32) -> [u8; M] {
```

Comment:
Actually, this is not necessarily the case that `end - start` and `M` are equal, which is why I put this condition in the loop. If you take the following code for example:
```rust
let db_mask_len = em_len - h_len - 1;
let offset = key_len - em_len;
let masked_db: [u8; 223] = get_array_slice(em, offset, db_mask_len + offset);
```
`db_mask_len` will vary as there will be trailing zeros for key sizes below 2048 bits, 223 being the value `db_mask_len` will take for a 2048-bit public key. So because the array has to be constant-sized `end - start` will be less than or equal to M. I added an assert to check this inequality though. And I added some details in the comments to explain the reasoning.

Also, I ended up increasing the size of the array to 479 to account for key sizes up to 4096 bits, as with 223 it will not work for anything above 2048 bits. But this increased the number of constraints substantially (+30k). Which brings me to something that would truly help in this kind of situations, being able to do operations with generics and constant numbers for the size of arrays like so:
```rust
let masked_db: [u8; NumBytes - 32 - 1] = get_array_slice(em, offset, db_mask_len + offset);
```
This would make the code much more efficient, but it's not possible right now even if I don't see any reason as to why this shouldn't be computable at compile time.

Reply:
Coming to think of it, the best trade-off would be to do like this in the meantime:
```rust
let masked_db: [u8; NumBytes] = get_array_slice(em, offset, db_mask_len + offset);
```
The array will always be 33 bytes too long, but that's better than having a fixed perfect size of 4096 bits that will add unnecessary constraints for most people as the most common size remains 2048 bits. Plus it brings more flexibility as the max size is no longer hard-coded.

Reply:
Alright, good to know that's supported now. I'm using Noir 0.32 on my machine at the moment. I assume you're referring to Noir 0.33. I've tried using Noir 0.33 before, but a few breaking changes broke the big int library, so I reverted back to Noir 0.32. But yes, it looks like the linter is not bugging me anymore when I try the code I mentioned with nargo 0.33. I'll try to fix the issues with the big int library (at least locally) to get it to compile and run the tests and I'll commit the changes.

Reply:
Just pushed it. You can have a look at the necessary fixes for the big int library here https://github.com/madztheo/noir-bignum/commit/d0c211d605ce12af3b2aab51317e9681189c3f2a

This means that the lib only works with Noir 0.33 and above though. But at the same time, the library doesn't work in its current state with Noir 0.33 because of the issues with the big int library due to the `to_le_bytes` function not taking any argument anymore.